myia/abstract/utils.py

```diff
@@ -195,20 +197,169 @@ def build_type_fn(self, x: Function):
 
 
+############
+# Checking #
+############
+
+
+@dataclass
+class CheckState:
+    """State of abstract_check."""
+
+    cache: dict
+    prop: str
+
+
+@overload.wrapper(initial_state=lambda: None)
+def abstract_check(__call__, self, x, *args):
+    """Check that the given object has a certain property."""
+    def proceed():
+        if hasattr(x, prop):
+            return getattr(x, prop) is x
+        elif __call__(self, x, *args):
+            if isinstance(x, (AbstractValue, Context, Reference)):
+                setattr(x, prop, x)
+            return True
+        else:
+            return False
+
+    prop = self.state.prop
+    cache = self.state.cache
+
+    try:
+        rval = cache.get(x, None)
+    except TypeError:
+        return proceed()
+
+    if rval is None:
+        cache[x] = True
+        cache[x] = proceed()
+        return cache[x]
+    else:
+        return rval
+
+
+@overload  # noqa: F811
+def abstract_check(self, x: TrackDict, *args):
+    return all(self(v, *args) for v in x.values())
+
+
+@overload  # noqa: F811
+def abstract_check(self, x: AbstractScalar, *args):
+    return self(x.values, *args)
+
+
+@overload  # noqa: F811
+def abstract_check(self, xs: AbstractStructure, *args):
+    return all(self(x, *args) for x in xs.children())
+
+
+@overload  # noqa: F811
+def abstract_check(self, xs: AbstractAtom, *args):
+    return True
+
+
+@overload  # noqa: F811
+def abstract_check(self, x: AbstractFunction, *args):
+    return self(x.values, *args)
+
+
+@overload  # noqa: F811
+def abstract_check(self, x: Possibilities, *args):
+    return all(self(v, *args) for v in x)
+
+
+@overload  # noqa: F811
+def abstract_check(self, t: PartialApplication, *args):
+    return self(t.fn, *args) and all(self(v, *args) for v in t.args)
+
+
+@overload  # noqa: F811
+def abstract_check(self, xs: object, *args):
+    return True
+
+
+@overload  # noqa: F811
+def abstract_check(self, ctx: Context, *args):
+    return (self(ctx.parent, *args)
+            and all(self(x, *args) for x in ctx.argkey))
+
+
+@overload  # noqa: F811
+def abstract_check(self, ref: Reference, *args):
+    return self(ref.context, *args)
+
+
 ###########
 # Cloning #
 ###########
 
 
-@overload(bootstrap=True)
-def abstract_clone(self, x: AbstractScalar, *args):
+@dataclass
+class CloneState:
+    """State of abstract_clone."""
+
+    cache: dict
+    prop: str
+    check: callable
+
+
+def _make_constructor(inst):
+    def f(*args, **kwargs):
+        inst.__init__(*args, **kwargs)
+        return inst
+    return f
+
+
+@overload.wrapper(
+    initial_state=lambda: CloneState({}, None, None)
+)
+def abstract_clone(__call__, self, x, *args):
     """Clone an abstract value."""
+    def proceed():
+        if isinstance(x, AbstractValue) and x in cache:
+            return cache[x]
+        result = __call__(self, x, *args)
+        if not isinstance(result, GeneratorType):
+            return result
+        constructor = None
+        while True:
```

Comment:
Why does this need to loop while the async version doesn't?

Reply:
Neither really "needs" to loop, the issue is that async generators can't return a value, so I have to treat the second yield as the return value, whereas normal generators raise `StopIteration` with the function's return value. I should probably remove the loop anyway, since it's not clear what entering it twice is supposed to mean.

Reply:
You should remove the loop from the non-async variant since it should only ever loop once and it would be clearer with an explicit second invocation of `send()`, possibly followed by an assert  that we don't continue past that.